### ISSUE TITLE
project does not compile with nannou_egui

### ISSUE THREAD

**Issue by elhmn** [tomoyanonymous/nannou-web-template#5]
Hi folks, and thank you for setting up that template ❤️.

I managed to get our [nannou-app](https://github.com/Git-Math/qautomata/pull/42) working on the browser. But things got a little complicated when I tried to add `nannou_egui` as a dependency to the project.

I only added nannou as a dependency on this [commit](https://github.com/Git-Math/qautomata/pull/43/commits/39d629c6968803dd6b86f33e9ef2a20dace29010) and the project failed to compile for the web as you can see on the error message bellow 👇 .  

```console
warning: /Users/elhmn/personal/qautomata/ui/Cargo.toml: Found `debug_assertions` in `target.'cfg(...)'.dependencies`. This value is not supported for selecting dependencies and will not work as expected. To learn more visit https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies
   Compiling winit v0.25.0
error: Please select a feature to build for web: `web-sys`, `stdweb`
  --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/mod.rs:38:1
   |
38 | compile_error!("Please select a feature to build for web: `web-sys`, `stdweb`");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

<details>

<summary>Expand: Full error dump after running `npm start`</summary>

```console
> rust-webpack-template@0.1.0 start
> rimraf dist pkg && webpack-dev-server --open

🧐  Checking for wasm-pack...

✅  wasm-pack is installed at /Users/elhmn/.cargo/bin/wasm-pack.

ℹ️  Compiling your crate in release mode...

<i> [webpack-dev-server] Project is running at:
<i> [webpack-dev-server] Loopback: http://localhost:8080/
<i> [webpack-dev-server] On Your Network (IPv4): http://192.168.178.82:8080/
<i> [webpack-dev-server] On Your Network (IPv6): http://[fe80::1]:8080/
<i> [webpack-dev-server] Content not from webpack is served from '/Users/elhmn/personal/qautomata/ui/dist' directory
<i> [webpack-dev-middleware] wait until bundle finished: /
[INFO]: 🎯  Checking for the Wasm target...
[INFO]: 🌀  Compiling to Wasm...
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/elhmn/personal/qautomata/core/Cargo.toml
workspace: /Users/elhmn/personal/qautomata/Cargo.toml
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/elhmn/personal/qautomata/ui/Cargo.toml
workspace: /Users/elhmn/personal/qautomata/Cargo.toml
warning: /Users/elhmn/personal/qautomata/ui/Cargo.toml: Found `debug_assertions` in `target.'cfg(...)'.dependencies`. This value is not supported for selecting dependencies and will not work as expected. To learn more visit https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies
   Compiling winit v0.25.0
error: Please select a feature to build for web: `web-sys`, `stdweb`
  --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/mod.rs:38:1
   |
38 | compile_error!("Please select a feature to build for web: `web-sys`, `stdweb`");
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0432]: unresolved import `super::backend`
 --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/event_loop/runner.rs:1:37
  |
1 | use super::{super::ScaleChangeArgs, backend, state::State};
  |                                     ^^^^^^^ no `backend` in `platform_impl::platform::event_loop`

error[E0432]: unresolved import `super::backend`
 --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/event_loop/state.rs:1:5
  |
1 | use super::backend;
  |     ^^^^^^^^^^^^^^ no `backend` in `platform_impl::platform::event_loop`

error[E0432]: unresolved import `super::backend`
 --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/event_loop/window_target.rs:1:29
  |
1 | use super::{super::monitor, backend, device, proxy::Proxy, runner, window};
  |                             ^^^^^^^ no `backend` in `platform_impl::platform::event_loop`

error[E0432]: unresolved import `super::backend`
 --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/event_loop/mod.rs:9:13
  |
9 | use super::{backend, device, window};
  |             ^^^^^^^ no `backend` in `platform_impl::platform`

error[E0432]: unresolved import `super::backend`
  --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/window.rs:12:13
   |
12 | use super::{backend, monitor, EventLoopWindowTarget};
   |             ^^^^^^^ no `backend` in `platform_impl::platform`

error[E0433]: failed to resolve: could not find `backend` in `super`
   --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/window.rs:156:16
    |
156 |         super::backend::scale_factor()
    |                ^^^^^^^ could not find `backend` in `super`

error[E0283]: type annotations needed
   --> /Users/elhmn/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.25.0/src/platform_impl/web/window.rs:349:5
    |
347 | #[derive(Default, Clone)]
    |          ------- in this derive macro expansion
348 | pub struct PlatformSpecificBuilderAttributes {
349 |     pub(crate) canvas: Option<backend::RawCanvasType>,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type
    |
    = note: cannot satisfy `_: Default`
    = note: this error originates in the derive macro `Default` (in Nightly builds, run with -Z macro-backtrace for more info)

Some errors have detailed explanations: E0283, E0432, E0433.
For more information about an error, try `rustc --explain E0283`.
error: could not compile `winit` due to 8 previous errors
Error: Compiling your crate to WebAssembly failed
Caused by: Compiling your crate to WebAssembly failed
Caused by: failed to execute `cargo build`: exited with exit status: 101
  full command: cd "/Users/elhmn/personal/qautomata/ui" && "cargo" "build" "--lib" "--release" "--target" "wasm32-unknown-unknown"
asset index.js 117 KiB [emitted] [minimized] (name: index) 1 related asset
asset index.html 293 bytes [emitted] [from: static/index.html] [copied]
asset 235.js 111 bytes [emitted] [minimized]
runtime modules 31.6 KiB 15 modules
orphan modules 18.8 KiB [orphan] 8 modules
cacheable modules 160 KiB
  modules by path ./node_modules/ 160 KiB
    modules by path ./node_modules/webpack-dev-server/client/ 55.8 KiB 4 modules
    modules by path ./node_modules/webpack/hot/*.js 4.59 KiB
      ./node_modules/webpack/hot/dev-server.js 1.88 KiB [built] [code generated]
      + 3 modules
    modules by path ./node_modules/html-entities/lib/*.js 81.3 KiB
      ./node_modules/html-entities/lib/index.js 7.74 KiB [built] [code generated]
      + 3 modules
    ./node_modules/ansi-html-community/index.js 4.16 KiB [built] [code generated]
    ./node_modules/events/events.js 14.5 KiB [built] [code generated]
  ./js/index.js 146 bytes [built] [code generated]
  ./pkg/index.js 1 bytes [built] [code generated]
webpack 5.78.0 compiled successfully in 1514 ms
```

</details>

I spent some time digging but I am very new to rust and to computer graphics, so I thought it would be great to ask for help here.
Maybe `nannou_egui` is not yet compatible with the experimental `wasm` branch of `nannou` and require some tunning/fixing ?

/cc @Git-Math

**Comment by tomoyanonymous** [comment]
Hi, thank you for reporting the problem.

AFAIK, the version of `egui` used in `nannou_egui` is quite old (0.15.0, the latest one is 0.21.0. egui is updating really fast...).
Several people (and also I) tried to updated the version of `egui` in `nannou_egui` but `egui` after  0.16.0 had a big change around wgpu backend. It's not easy to update `nannou_egui` without touching the overall backend of nannou.

Please refer this open issue https://github.com/nannou-org/nannou/issues/837 and the working PR to solve it https://github.com/nannou-org/nannou/pull/861

In other words, nannou and egui each have their own native/web compatibility support, and that the conflict between the two is occurring in nannou-egui.

Honestly, that is one of the reason that I started my project with `nannou_egui`, moved to the combination of `nannou` & pure `egui` and finally `egui` & `eframe`,the built-in app framework in egui after making this template repo.

**Comment by elhmn** [comment]
Thanks for the answer and the tracking issues, I will follow the progress of the update there.

> Honestly, that is one of the reason that I started my project with `nannou_egui`, moved to the combination of `nannou` & pure `egui` and finally `egui` & `eframe`,the built-in app framework in egui after making this template repo.

Is the new project you started working on is open source ? I am interested in knowing how you the setup with `egui` & `eframe` and `nannou` looks like.

**Comment by tomoyanonymous** [comment]
Yes, I worked on a separate branch to remove nannou as a dependency, you can track how I changed the code.

https://github.com/tomoyanonymous/otopoiesis/commit/d5ff18a4f71be6ed3f0e6038ba06f341e3a4caa3

**Comment by elhmn** [comment]
> Yes, I worked on a separate branch to remove nannou as a dependency, you can track how I changed the code.
> 
> 
> 
> https://github.com/tomoyanonymous/otopoiesis/commit/d5ff18a4f71be6ed3f0e6038ba06f341e3a4caa3

Awesome thank you 🙇‍♂️
